cpp/wallet_server.cc

```diff
@@ -0,0 +1,344 @@
+    stats_context.AddMetadata("membership", membership_);
+    PriceRequest stats_request;
+    PriceResponse stats_response;
+    // Call Stats Server to fetch the price to calculate the balance.
+    Status stats_status =
+        stats_stub_->FetchPrice(&stats_context, stats_request, &stats_response);
+    if (stats_status.ok()) {
+      auto metadata_hostname =
+          stats_context.GetServerInitialMetadata().find("hostname");
+      if (metadata_hostname != stats_context.GetServerInitialMetadata().end()) {
+        std::cout << "server host: "
+                  << std::string(metadata_hostname->second.data(),
+                                 metadata_hostname->second.length())
+                  << std::endl;
+      }
+      std::cout << "grpc-coin price " << stats_response.price() << std::endl;
+    } else {
+      std::cout << stats_status.error_code() << ": "
+                << stats_status.error_message() << std::endl;
+    }
+    int total_per_user = 0;
+    for (auto& address_entry : user_coin_map_[user_]) {
+      if (!v1_behavior_ && request->include_balance_per_address()) {
+        auto* address = response->add_addresses();
+        address->set_address(address_entry.first);
+        address->set_balance(address_entry.second * stats_response.price());
+      }
+      total_per_user += address_entry.second;
+    }
+    response->set_balance(total_per_user * stats_response.price());
+    return Status::OK;
+  }
+
+  Status WatchBalance(ServerContext* context, const BalanceRequest* request,
+                      ServerWriter<BalanceResponse>* writer) override {
+    if (!ObtainAndValidateUserAndMembership(context)) {
+      return Status(StatusCode::UNAUTHENTICATED,
+                    "membership authentication failed");
+    }
+    context->AddInitialMetadata("hostname", hostname_);
+    ClientContext stats_context;
+    stats_context.AddMetadata("authorization", token_);
+    stats_context.AddMetadata("membership", membership_);
+    PriceRequest stats_request;
+    PriceResponse stats_response;
+    // Open a streaming price watching with Stats Server.
+    // Every time a response
+    // is received, use the price to calculate the balance.
+    // Send every updated balance in a stream back to the client.
+    std::unique_ptr<ClientReader<PriceResponse>> stats_reader(
+        stats_stub_->WatchPrice(&stats_context, stats_request));
+    while (stats_reader->Read(&stats_response)) {
+      auto metadata_hostname =
+          stats_context.GetServerInitialMetadata().find("hostname");
+      if (metadata_hostname != stats_context.GetServerInitialMetadata().end()) {
+        std::cout << "server host: "
+                  << std::string(metadata_hostname->second.data(),
+                                 metadata_hostname->second.length())
+                  << std::endl;
+      }
+      std::cout << "grpc-coin price: " << stats_response.price() << std::endl;
+      BalanceResponse response;
+      int total_per_user = 0;
```

Comment:
Make a function for this `for { total += per_address }`?

Reply:
done